### PR TITLE
[improvement](scan) Support pushdown execute expr ctx

### PR DESCRIPTION
**Pull request by xinyiZzz**
# Proposed changes

Issue Number: close #xxx

#### motivation
In the past, only simple predicates (`slot=const`), `and`, `like`, `or` (only bitmap index) could be pushed down to the storage layer. scan process:
1. Read part of the column first, and calculate the row ids with a simple push-down predicate.
2. Use row ids to read the remaining columns and pass them to the scanner, and the scanner filters the remaining predicates.

This pr will also push-down the remaining predicates (functions, nested predicates...) in the scanner to the storage layer for filtering. scan process:
1. Read part of the column first, and use the push-down simple predicate to calculate the row ids, (same as above)
2. Use row ids to read the columns needed for the remaining predicates, and use the pushed-down remaining predicates to reduce the number of row ids again.
3. Use row ids to read the remaining columns and pass them to the scanner.

#### expected
1. `select a,b,c,... from tbl where func(a) = 1`. The performance will be significantly improved, and the effect depends on the number of selected columns and the filtering rate.
    
example:  10 times better performance
```
SELECT * FROM hits WHERE UPPER(URL)="HTTP://SAMARA.IRR.RU/CATALOG_GOOGLETBR%26AD%3D278885%26BT%3D430001216";
```
2. `select a from tbl where func(a) = 1`. Performance is expected to remain the same, as it behaves the same as before. 
(In the test, there is a 2% loss in performance. I found that `VExprContext::filter_block` performs differently in different locations, which may be related to cache miss, but the tool did not detect it)

1. clickbench test

|    test    | cost(s) |
| ---------- | --- |
| default | 134 |
| set enable_remaining_expr_pushdown = true; |  110.25 |

default:
```
query1: 0.03,0.03,0.03
query2: 0.11,0.04,0.04
query3: 0.09,0.07,0.07
query4: 0.22,0.08,0.07
query5: 0.66,0.62,0.54
query6: 0.97,0.90,0.93
query7: 0.02,0.01,0.01
query8: 0.05,0.05,0.04
query9: 1.55,1.38,1.41
query10: 1.98,1.93,1.94
query11: 0.36,0.33,0.35
query12: 0.37,0.34,0.38
query13: 0.61,0.63,0.59
query14: 1.46,1.46,1.44
query15: 0.84,0.85,0.88
query16: 0.51,0.48,0.47
query17: 1.78,1.96,1.82
query18: 0.45,0.43,0.42
query19: 4.25,4.25,4.19
query20: 0.01,0.02,0.01
query21: 1.20,0.31,0.31
query22: 0.12,0.11,0.11
query23: 1.13,0.29,0.28
query24: 8.33,9.99,9.83
query25: 0.13,0.10,0.11
query26: 0.10,0.10,0.11
query27: 0.11,0.09,0.11
query28: 0.48,0.37,0.38
query29: 4.74,4.63,4.64
query30: 1.95,1.98,1.91
query31: 0.43,0.38,0.41
query32: 0.61,0.59,0.58
query33: 4.04,4.03,3.94
query34: 4.52,3.86,3.94
query35: 4.07,3.63,3.79
query36: 1.41,1.43,1.37
query37: 0.07,0.06,0.05
query38: 0.03,0.03,0.03
query39: 0.02,0.02,0.03
query40: 0.15,0.13,0.14
query41: 0.04,0.02,0.03
query42: 0.03,0.02,0.02
query43: 0.04,0.04,0.03
```

set enable_remaining_expr_pushdown = true;
```
query1: 0.02,0.02,0.01
query2: 0.03,0.04,0.03
query3: 0.06,0.07,0.07
query4: 0.07,0.07,0.07
query5: 0.67,0.60,0.56
query6: 0.87,0.84,0.84
query7: 0.01,0.01,0.01
query8: 0.05,0.03,0.04
query9: 1.48,1.36,1.41
query10: 2.02,1.88,1.90
query11: 0.32,0.44,0.35
query12: 0.35,0.34,0.42
query13: 0.59,0.70,0.64
query14: 1.46,1.48,1.44
query15: 0.81,0.92,0.82
query16: 0.49,0.46,0.46
query17: 1.98,1.98,1.92
query18: 0.43,0.42,0.44
query19: 4.31,4.29,4.38
query20: 0.01,0.01,0.01
query21: 0.30,0.31,0.32
query22: 0.11,0.12,0.11
query23: 0.26,0.27,0.26
query24: 0.44,0.44,0.43
query25: 0.11,0.10,0.09
query26: 0.11,0.11,0.11
query27: 0.10,0.10,0.11
query28: 0.37,0.38,0.39
query29: 4.71,4.59,4.62
query30: 2.04,2.04,1.95
query31: 0.43,0.39,0.38
query32: 0.61,0.63,0.58
query33: 3.98,4.08,3.93
query34: 4.21,3.97,4.08
query35: 3.89,3.98,3.56
query36: 1.36,1.32,1.35
query37: 0.06,0.05,0.06
query38: 0.03,0.02,0.03
query39: 0.02,0.02,0.02
query40: 0.15,0.14,0.14
query41: 0.02,0.02,0.02
query42: 0.02,0.02,0.02
query43: 0.04,0.03,0.02
```

## Checklist(Required)

1. Does it affect the original behavior: 
    - [ ] Yes
    - [ ] No
    - [ ] I don't know
4. Has unit tests been added:
    - [ ] Yes
    - [ ] No
    - [ ] No Need
5. Has document been added or modified:
    - [ ] Yes
    - [ ] No
    - [ ] No Need
6. Does it need to update dependencies:
    - [ ] Yes
    - [ ] No
7. Are there any changes that cannot be rolled back:
    - [ ] Yes (If Yes, please explain WHY)
    - [ ] No

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: be/src/olap/schema.h [hunk 1]
```diff
@@ -144,6 +144,7 @@ class Schema {
     const std::vector<ColumnId>& column_ids() const { return _col_ids; }
     const std::vector<int32_t>& unique_ids() const { return _unique_ids; }
     ColumnId column_id(size_t index) const { return _col_ids[index]; }
+    ColumnId column_id(const std::string& cname) const { return _field_name_to_col_id.at(cname); }
```
Comment (yiguolei) [inline be/src/olap/schema.h:147]: why need column name here?  column name is not right during light weight schema change. For example, user call alter table rename column

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 35.77 seconds
 load time: 491 seconds
 storage size: 17122669491 Bytes
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20230116024129_clickbench_pr_80700.html

File: be/src/olap/schema.h (hunk 1 above)
Comment (xinyiZzz) [reply be/src/olap/schema.h:147]: fixed

**Comment by xinyiZzz** [comment]
run p0

**Comment by xinyiZzz** [comment]
run buildall

**Comment by xinyiZzz** [comment]
run arm

**Comment by xinyiZzz** [comment]
run p0

**Comment by xinyiZzz** [comment]
run buildall

**Comment by xinyiZzz** [comment]
run buildall

**Comment by xinyiZzz** [comment]
run clickbench

**Comment by xinyiZzz** [comment]
run clickbench

File: be/src/vec/core/block.h [hunk 2]
```diff
@@ -75,6 +75,7 @@ class Block {
     IndexByName index_by_name;
     std::vector<bool> row_same_bit;
```
Comment (yiguolei) [inline be/src/vec/core/block.h:77]: Add some comment here to describe why need a _effective_col

File: be/src/runtime/runtime_state.h [hunk 3]
```diff
@@ -118,6 +118,11 @@ class RuntimeState {
                _query_options.check_overflow_for_decimal;
     }
 
+    bool enable_remaining_expr_pushdown() const {
```
Comment (yiguolei) [inline be/src/runtime/runtime_state.h:121]: use another name, for example: enable_common_expr_pushdown

File: be/src/vec/olap/vcollect_iterator.cpp [hunk 4]
```diff
@@ -285,11 +285,6 @@ Status VCollectIterator::_topn_next(Block* block) {
 
             auto col_name = block->get_names()[first_sort_column_idx];
 
-            // filter block
-            RETURN_IF_ERROR(VExprContext::filter_block(
```
Comment (yiguolei) [inline be/src/vec/olap/vcollect_iterator.cpp:289]: Do you check this with hangyu? it is useless?

File: be/src/vec/exec/scan/vscanner.cpp [hunk 5]
```diff
@@ -67,12 +67,12 @@ Status VScanner::get_block(RuntimeState* state, Block* block, bool* eof) {
             }
 
             // 2. Filter the output block finally.
-            {
+            if (!_enable_remaining_expr_pushdown) {
                 SCOPED_TIMER(_parent->_filter_timer);
```
Comment (yiguolei) [inline be/src/vec/exec/scan/vscanner.cpp:71]: Maybe a problem here. Runtime filter comes after scanner start. Then runtime filter will never take affect.

File: be/src/olap/rowset/segment_v2/segment_iterator.cpp [hunk 6]
```diff
@@ -31,9 +31,18 @@
 #include "util/doris_metrics.h"
 #include "util/key_util.h"
 #include "util/simd/bits.h"
+#include "vec/columns/column.h"
+// #include "vec/columns/column_array.h"
+#include "vec/columns/column_const.h"
+// #include "vec/columns/column_nullable.h"
```
Comment (yiguolei) [inline be/src/olap/rowset/segment_v2/segment_iterator.cpp:37]: remove these comment.

File: be/src/olap/rowset/segment_v2/segment_iterator.cpp [hunk 7]
```diff
@@ -181,7 +190,10 @@ Status SegmentIterator::init(const StorageReadOptions& opts) {
         _output_columns = *(opts.output_columns);
     }
 
-    _remaining_vconjunct_root = opts.remaining_vconjunct_root;
+    _remaining_vconjunct_ctx = opts.remaining_vconjunct_ctx;
```
Comment (yiguolei) [inline be/src/olap/rowset/segment_v2/segment_iterator.cpp:193]: any difference between _remaining_vconjunct_ctx and _remaining_vconjunct_root?

File: be/src/vec/core/block.h (hunk 2 above)
Comment (xinyiZzz) [reply be/src/vec/core/block.h:77]: fixed, add `set_effective_col` comment

File: be/src/vec/olap/vcollect_iterator.cpp (hunk 4 above)
Comment (xinyiZzz) [reply be/src/vec/olap/vcollect_iterator.cpp:289]: Checked with them, can be replaced

File: be/src/olap/rowset/segment_v2/segment_iterator.cpp (hunk 6 above)
Comment (xinyiZzz) [reply be/src/olap/rowset/segment_v2/segment_iterator.cpp:37]: fixed

File: be/src/olap/rowset/segment_v2/segment_iterator.cpp (hunk 7 above)
Comment (xinyiZzz) [reply be/src/olap/rowset/segment_v2/segment_iterator.cpp:193]: `_remaining_vconjunct_root` = `_remaining_vconjunct_ctx->root()`
Got `_remaining_vconjunct_root` in `SegmentIterator::init`. Compatible before

File: be/src/olap/rowset/segment_v2/segment_iterator.cpp [hunk 8]
```diff
@@ -1160,58 +1189,86 @@ void SegmentIterator::_vec_init_lazy_materialization() {
...
+                    _non_predicate_columns.push_back(cid);
+                } else {
+                    _second_read_column_ids.push_back(cid);
+                }
                 if (_is_need_vec_eval || _is_need_short_eval) {
                     _lazy_materialization_read = true;
                 }
             }
         }
     }
```
Comment (yiguolei) [inline be/src/olap/rowset/segment_v2/segment_iterator.cpp:1230]: there are  3 if else here. I think the code is right, but it too hard to think it correctly. 
Better add a comment here.

File: be/src/olap/rowset/segment_v2/segment_iterator.cpp [hunk 9]
```diff
@@ -1160,58 +1189,86 @@ void SegmentIterator::_vec_init_lazy_materialization() {
...
+                    _is_pred_column[cid] = true;
+                }
+            }
+        } else if (_is_need_expr_eval) {
+            DCHECK(!_is_need_vec_eval && !_is_need_short_eval);
+            for (auto cid : _remaining_expr_columns) {
                 _first_read_column_ids.push_back(cid);
-                // when _lazy_materialization_read = false, non-predicate column should also be filtered by sel idx, so we regard it as pred columns
-                _is_pred_column[cid] = true;
             }
```
Comment (yiguolei) [inline be/src/olap/rowset/segment_v2/segment_iterator.cpp:1269]: in this case, should clear _second_read_column_ids?

File: be/src/olap/rowset/segment_v2/segment_iterator.cpp [hunk 10]
```diff
@@ -1505,14 +1562,16 @@ Status SegmentIterator::next_batch(vectorized::Block* block) {
 
     _current_batch_rows_read = 0;
     uint32_t nrows_read_limit = _opts.block_row_max;
-    if (UNLIKELY(_estimate_row_size)) {
-        // read 100 rows to estimate average row size
+    if (UNLIKELY(_opts.stats->blocks_load == 0)) {
```
Comment (yiguolei) [inline be/src/olap/rowset/segment_v2/segment_iterator.cpp:1565]: Not depend on the stats, add another variable to indicate that this is first read.

File: be/src/vec/exec/scan/vscanner.cpp (hunk 5 above)
Comment (xinyiZzz) [reply be/src/vec/exec/scan/vscanner.cpp:71]: Copy `_vconjunct_ctx_ptr` to `_common_vexpr_ctxs_pushdown` and push down, mark `_vconjunct_ctx_ptr` as stale

File: be/src/olap/rowset/segment_v2/segment_iterator.cpp (hunk 8 above)
Comment (xinyiZzz) [reply be/src/olap/rowset/segment_v2/segment_iterator.cpp:1230]: fixed

File: be/src/olap/rowset/segment_v2/segment_iterator.cpp (hunk 9 above)
Comment (xinyiZzz) [reply be/src/olap/rowset/segment_v2/segment_iterator.cpp:1269]: In this case `_second_read_column_ids` is empty and there is no common expr pushdown.
Added a comment

File: be/src/olap/rowset/segment_v2/segment_iterator.cpp (hunk 10 above)
Comment (xinyiZzz) [reply be/src/olap/rowset/segment_v2/segment_iterator.cpp:1565]: fixed
// Read up to 100 rows at a time while waiting for the estimated row size.

**Comment by xinyiZzz** [comment]
run buildall

**Comment by xinyiZzz** [comment]
run buildall

**Comment by xinyiZzz** [comment]
run buildall

File: be/src/vec/exec/scan/vscan_node.cpp [hunk 11]
```diff
@@ -458,7 +461,15 @@ Status VScanNode::_normalize_conjuncts() {
             RETURN_IF_ERROR(_normalize_predicate((*_vconjunct_ctx_ptr)->root(), &new_root));
             if (new_root) {
                 (*_vconjunct_ctx_ptr)->set_root(new_root);
-            } else {
+                if (_should_push_down_common_expr()) {
+                    // Copy _vconjunct_ctx_ptr to _common_vexpr_ctxs_pushdown and push down, mark _vconjunct_ctx_ptr as stale.
```
Comment (yiguolei) [inline be/src/vec/exec/scan/vscan_node.cpp:465]: 这里不需要这么写； 我感觉只要把两个指针交换一下，然后把_vconjunct_ctx_ptr 设置为nullptr 就行了

Comment (xinyiZzz) [reply be/src/vec/exec/scan/vscan_node.cpp]: done

**Comment by xinyiZzz** [comment]
run buildall

**Comment by xinyiZzz** [comment]
run buildall

**Comment by k-i-d-d** [comment]
run beut macos

**Comment by xinyiZzz** [comment]
run buildall

File: be/src/vec/exec/scan/vscan_node.cpp [hunk 12]
```diff
@@ -458,7 +461,11 @@ Status VScanNode::_normalize_conjuncts() {
             RETURN_IF_ERROR(_normalize_predicate((*_vconjunct_ctx_ptr)->root(), &new_root));
             if (new_root) {
                 (*_vconjunct_ctx_ptr)->set_root(new_root);
-            } else {
+                if (_should_push_down_common_expr()) {
+                    _common_vexpr_ctxs_pushdown = *_vconjunct_ctx_ptr;
```
Comment (yiguolei) [inline be/src/vec/exec/scan/vscan_node.cpp:465]: maybe memory leak here

File: be/src/vec/core/block.cpp [hunk 13]
```diff
@@ -321,13 +322,16 @@ void Block::check_number_of_rows(bool allow_null_columns) const {
...
-            LOG(FATAL) << fmt::format("Sizes of columns doesn't match: {}:{},{}:{}",
-                                      data.front().name, rows, elem.name, size);
+            LOG(FATAL) << fmt::format("Sizes of columns doesn't match: {}:{},{}:{}, col size: {}",
+                                      data.front().name, rows, elem.name, size, each_col_size());
         }
     }
 }
 
 size_t Block::rows() const {
+    if (_effective_col != INT_MIN) {
```
Comment (yiguolei) [inline be/src/vec/core/block.cpp:332]: use column const during filter block. and then replace column after filter block.
Then do not need add this effective col any more.

File: fe/fe-core/src/main/java/org/apache/doris/qe/SessionVariable.java [hunk 14]
```diff
@@ -642,6 +644,9 @@ public class SessionVariable implements Serializable, Writable {
     @VariableMgr.VarAttr(name = ENABLE_FUNCTION_PUSHDOWN)
     public boolean enableFunctionPushdown = true;
 
+    @VariableMgr.VarAttr(name = ENABLE_COMMON_EXPR_PUSHDOWN, fuzzy = true)
+    public boolean enableCommonExprPushdown = true;
```
Comment (yiguolei) [inline fe/fe-core/src/main/java/org/apache/doris/qe/SessionVariable.java:648]: should add fuzzy code logic in initFuzzyModeVariables method.

File: be/src/olap/rowset/segment_v2/segment_iterator.cpp [hunk 15]
```diff
@@ -1196,58 +1220,91 @@ void SegmentIterator::_vec_init_lazy_materialization() {
...
+            // pred exits, but we can eliminate lazy materialization
+            // insert pred/non-pred cid to first read columns
+            std::set<ColumnId> pred_id_set;
+            pred_id_set.insert(_short_cir_pred_column_ids.begin(),
+                               _short_cir_pred_column_ids.end());
+            pred_id_set.insert(_vec_pred_column_ids.begin(), _vec_pred_column_ids.end());
+            std::set<ColumnId> non_pred_set(_non_predicate_columns.begin(),
+                                            _non_predicate_columns.end());
+
+            // _second_read_column_ids must be empty. Otherwise _lazy_materialization_read must not false.
```
Comment (yiguolei) [inline be/src/olap/rowset/segment_v2/segment_iterator.cpp:1290]: ADD CHECK(!_second_read_column_ids.empty())

File: fe/fe-core/src/main/java/org/apache/doris/qe/SessionVariable.java (hunk 14 above)
Comment (xinyiZzz) [reply fe/fe-core/src/main/java/org/apache/doris/qe/SessionVariable.java:648]: done

File: be/src/vec/core/block.cpp (hunk 13 above)
Comment (xinyiZzz) [reply be/src/vec/core/block.cpp:332]: done

**Comment by xinyiZzz** [comment]
run buildall

**Review by yiguolei — DISMISSED** [review]
LGTM

**Comment by yiguolei** [comment]
./run buildall

**Comment by xinyiZzz** [comment]
./run buildall

**Comment by xinyiZzz** [comment]
./run buildall

**Comment by xinyiZzz** [comment]
run buildall

**Comment by xinyiZzz** [comment]
run buildall

**Comment by xinyiZzz** [comment]
run buildall

**Review by yiguolei — APPROVED** [review]
LGTM

**Pull request merged**